File: edukate-notifier/src/main/java/io/github/sanyavertolet/edukate/notifier/entities/SimpleNotification.java
package io.github.sanyavertolet.edukate.notifier.entities;

import com.fasterxml.jackson.annotation.JsonTypeName;
import io.github.sanyavertolet.edukate.common.notifications.SimpleNotificationCreateRequest;
import io.github.sanyavertolet.edukate.notifier.dtos.SimpleNotificationDto;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.springframework.data.annotation.TypeAlias;

@Getter
@Setter
@NoArgsConstructor
@TypeAlias("simple")
@JsonTypeName("simple")
public final class SimpleNotification extends BaseNotification {
    private String title;
    private String message;
    private String source;

    public SimpleNotification(String uuid, String targetUserId, String title, String message, String source) {
        super(uuid, targetUserId);
        this.title = title;
        this.message = message;
        this.source = source;
    }

    @Override
    public SimpleNotificationDto toDto() {
        return new SimpleNotificationDto(getUuid(), getIsRead(), getCreatedAt(), title, message, source);
    }

    public static SimpleNotification fromCreationRequest(SimpleNotificationCreateRequest creationRequest) {
        return new SimpleNotification(
                creationRequest.getUuid(), creationRequest.getTargetUserId(),
                creationRequest.getTitle(), creationRequest.getMessage(), creationRequest.getSource()
        );
    }
}

File: edukate-notifier/src/main/kotlin/io/github/sanyavertolet/edukate/notifier/entities/SimpleNotification.kt
package io.github.sanyavertolet.edukate.notifier.entities

import com.fasterxml.jackson.annotation.JsonTypeName
import io.github.sanyavertolet.edukate.common.notifications.SimpleNotificationCreateRequest
import io.github.sanyavertolet.edukate.notifier.dtos.SimpleNotificationDto
import java.time.Instant
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.index.Indexed

@JsonTypeName("simple")
data class SimpleNotification(
    @field:Id override val id: String? = null,
    @field:Indexed(unique = true) override val uuid: String,
    override val targetUserId: Long,
    override val isRead: Boolean = false,
    @field:CreatedDate override val createdAt: Instant? = null,
    val title: String,
    val message: String,
    val source: String,
) : BaseNotification() {
    override fun toDto() =
        SimpleNotificationDto(
            uuid = uuid,
            isRead = isRead,
            createdAt = requireNotNull(createdAt) { "createdAt is null for notification uuid=$uuid" },
            title = title,
            message = message,
            source = source,
        )

    override fun markAsRead() = copy(isRead = true)

    companion object {
        @JvmStatic
        fun fromCreationRequest(creationRequest: SimpleNotificationCreateRequest) =
            SimpleNotification(
                uuid = creationRequest.uuid,
                targetUserId = creationRequest.targetUserId,
                title = creationRequest.title,
                message = creationRequest.message,
                source = creationRequest.source,
            )
    }
}
